mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +228,682 @@ def _chain_with_retriever(model):
+            **kwargs: Any,
+        ):
+            """Run on a retry event."""
+            span = self._get_span_by_run_id(run_id)
+            retry_d: Dict[str, Any] = {
+                "slept": retry_state.idle_for,
+                "attempt": retry_state.attempt_number,
+            }
+            if retry_state.outcome is None:
+                retry_d["outcome"] = "N/A"
+            elif retry_state.outcome.failed:
+                retry_d["outcome"] = "failed"
+                exception = retry_state.outcome.exception()
+                retry_d["exception"] = str(exception)
+                retry_d["exception_type"] = exception.__class__.__name__
+            else:
+                retry_d["outcome"] = "success"
+                retry_d["result"] = str(retry_state.outcome.result())
+            span.add_event(
+                SpanEvent(
+                    name="retry",
+                    timestamp=int(datetime.now(timezone.utc).timestamp()),
+                    attributes=retry_d,
+                )
+            )
+
+        @override
+        def on_llm_end(self, response: LLMResult, *, run_id: UUID, **kwargs: Any):
+            """End the span for an LLM run."""
+            llm_span = self._get_span_by_run_id(run_id)
+            outputs = response.dict()
+            for i, generations in enumerate(response.generations):
+                for j, generation in enumerate(generations):
+                    output_generation = outputs["generations"][i][j]
+                    if "message" in output_generation:
+                        output_generation["message"] = dumpd(
+                            cast(ChatGeneration, generation).message
+                        )
+            llm_span.add_event(
+                SpanEvent("end", timestamp=int(datetime.now(timezone.utc).timestamp()))
+            )
+            self._end_span(llm_span, outputs=outputs)
+            for i, generations in enumerate(response.generations):
+                for j, generation in enumerate(generations):
+                    text_complexity_metrics = self._analyze_text(generation.text) or {}
+                    self._log_artifact(
+                        {"generation": generation.dict(), **text_complexity_metrics},
+                        f"llm_end_{llm_span.span_id}_generation_{i}_{j}",
+                    )
+            self.metrics["step"] += 1
+            self.metrics["llm_ends"] += 1
+            self.metrics["ends"] += 1
+
+        @override
+        def on_llm_error(
+            self,
+            error: BaseException,
+            *,
+            run_id: UUID,
+            **kwargs: Any,
+        ):
+            """Handle an error for an LLM run."""
+            llm_span = self._get_span_by_run_id(run_id)
+            llm_span.add_event(
+                SpanEvent(
+                    "error",
+                    timestamp=int(datetime.now(timezone.utc).timestamp()),
+                    attributes={"error": str(error)},
```

Comment:
Would it be possible to use the standard format of OTel exception event? I think that would be helpful for the UI side to render errors nicely.
ref: https://github.com/mlflow/mlflow/blob/tracing/tests/entities/test_trace_data.py#L114-L119

Reply:
In that case I would prefer us defining a class named ErrorEvent and follow this pattern, otherwise it's impossible for users to know this and UI won't work so well?

Reply:
True, but we don't know how many users need to implement custom callbacks. If it's small, a few code examples might be enough. We can come back to this point once error UI design is fixed. Thanks!